R/zchunk_LA102.en_emiss_CDIAC.R

```diff
@@ -1,97 +1,147 @@
+      L1011.en_bal_EJ_R_Si_Fi_Yh
 
     # ===================================================
     # TRANSLATED PROCESSING CODE GOES HERE...
-    #
-    # If you find a mistake/thing to update in the old code and
-    # fixing it will change the output data, causing the tests to fail,
-    # (i) open an issue on GitHub, (ii) consult with colleagues, and
-    # then (iii) code a fix:
-    #
-    # if(OLD_DATA_SYSTEM_BEHAVIOR) {
-    #   ... code that replicates old, incorrect behavior
-    # } else {
-    #   ... new code with a fix
-    # }
-    #
-    #
-    # NOTE: there are 'match' calls in this code. You probably want to use left_join_error_no_match
-    # For more information, see https://github.com/JGCRI/gcamdata/wiki/Name-That-Function
-    # NOTE: This code uses vecpaste
-    # This function can be removed; see https://github.com/JGCRI/gcamdata/wiki/Name-That-Function
+
+    # 2. Perform computations
+    # prep region IDs for matching
+    iso_GCAM_regID %>%
+      select(iso, GCAM_region_ID) ->
+      iso_GCAM_regID_reg
+
+    # append GCAM region IDs to isos and map GCAM fuels to CDIAC fuels
+    L100.CDIAC_CO2_ctry_hist %>%
+      gather(fuel, value, -iso, -year) %>%
+      left_join_error_no_match(iso_GCAM_regID_reg, by = "iso") %>%
+      rename(CDIAC_fuel = `fuel`) %>%
+      left_join_error_no_match(CDIAC_fuel, by = "CDIAC_fuel") %>%
+
+    # Aggregate CO2 emissions by GCAM region and fuel
+      group_by(GCAM_region_ID, fuel, year) %>%
+      summarise(value = sum(value) * CONV_KT_MT) ->
+      L102.CO2_Mt_R_F_Yh
+
+    # Calculate regional and global CO2 emissions coefficients by fuel
+    # Calculate the TPES by fuel, deducting non-energy use of fuels that does not result in CO2 emissions
+
+    L1011.en_bal_EJ_R_Si_Fi_Yh %>%
+      filter(sector == "in_industry_feedstocks" & L1011.en_bal_EJ_R_Si_Fi_Yh$fuel %in% L102.CO2_Mt_R_F_Yh$fuel) %>%
+      left_join_error_no_match(A32.nonenergy_Cseq, by = c("fuel" = "subsector")) %>%
+      mutate(val = value * remove.fraction) %>%
+      select(GCAM_region_ID, fuel, year, val) ->
+      L102.en_sequestered_EJ_R_Fi_Yh
+
+    L1011.en_bal_EJ_R_Si_Fi_Yh %>%
+      filter(sector == "TPES" & fuel %in% L102.CO2_Mt_R_F_Yh$fuel) %>%
+      left_join_error_no_match(L102.en_sequestered_EJ_R_Fi_Yh, by = c("GCAM_region_ID", "fuel", "year")) %>%
+      mutate(val = value - val) %>%
+      select(-sector, -value) ->
+      L102.en_emitted_EJ_R_Fi_Yh
+
+    # Calculate the emissions coefficients by fuel, using only the energy whose carbon is assumed to be emitted
+    # regional
+    L102.CO2_Mt_R_F_Yh %>%
+      filter(fuel %in% L102.en_emitted_EJ_R_Fi_Yh$fuel) %>%
+      left_join_error_no_match(L102.en_emitted_EJ_R_Fi_Yh, by = c("GCAM_region_ID", "fuel", "year")) %>%
+      mutate(value = value / val) %>%
+      select(-val) %>%
+
+    # reset to defaults wherever NAs result from 0 energy consumption
+      mutate(value = if_else(fuel == "gas" & is.na(value), energy.DEFAULT_GAS_CCOEF, value)) %>%
```

Comment:
Is this any more efficient? It's the only easy way I could see to apply replace_na to select rows.

`    L102.CO2_Mt_R_F_Yh %>%
      filter(fuel == "gas") %>% 
      replace_na(value = DEFAULT_GAS_CCOEF) -> L102.CO2_Mt_R_F_Yh_gas
    
    L102.CO2_Mt_R_F_Yh %>%
      filter(fuel == "coal") %>% 
      replace_na(value = DEFAULT_COAL_CCOEF) -> L102.CO2_Mt_R_F_Yh_coal
    
    L102.CO2_Mt_R_F_Yh %>%
      filter(fuel == "refined liquids") %>% 
      replace_na(value = DEFAULT_LIQUIDS_CCOEF) %>% 
      bind_rows(L102.CO2_Mt_R_F_Yh_coal) %>% 
      bind_rows(L102.CO2_Mt_R_F_Yh_gas) ->
      L102.Ccoef_kgCGJ_R_F_Yh`